### PR TITLE
[Cache] Add a Chain adapter

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | no |
| Fixed tickets | #17556 |
| License | MIT |
| Doc PR | - |

Made in coordination with @dunglas

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
fabbot failure is a false positive (and a fabbot bug)

File: src/Symfony/Component/Cache/Adapter/AdapterInterface.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Cache\Adapter;
+
+use Psr\Cache\CacheItemPoolInterface;
+
+/**
+ * Marker interface for adapters managing {@see \Symfony\Component\Cache\CacheItem} instances.
```
Comment (javiereguiluz) [inline src/Symfony/Component/Cache/Adapter/AdapterInterface.php]: I can't understand this description:

```
Marker interface for adapters managing instances.
```

File: src/Symfony/Component/Cache/Adapter/ChainAdapter.php [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+ */
+
+namespace Symfony\Component\Cache\Adapter;
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\Cache\CacheItem;
+use Symfony\Component\Cache\Exception\InvalidArgumentException;
+
+/**
```
Comment (javiereguiluz) [inline src/Symfony/Component/Cache/Adapter/ChainAdapter.php]: The PHPdoc of this class could also be reworded a bit. It's confusing.

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/ChainAdapter.php]: updated

File: src/Symfony/Component/Cache/Adapter/ChainAdapter.php [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    public function getItem($key)
+    {
+        $saveUp = $this->saveUp;
+
+        foreach ($this->adapters as $i => $adapter) {
+            $item = $adapter->getItem($key);
+
+            if ($item->isHit()) {
+                while (0 <= --$i) {
+                    $saveUp($this->adapters[$i], $item);
```
Comment (dupuchba) [inline src/Symfony/Component/Cache/Adapter/ChainAdapter.php]: @nicolas-grekas I don't understand why you save the item with a new lifetime for a fetch operation. Can you explain it to me plz ?

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/ChainAdapter.php]: because there is no way to know the remaining lifetime of the item in its original pool...

Comment (dupuchba) [inline src/Symfony/Component/Cache/Adapter/ChainAdapter.php]: Ok, thx

**Comment by nicolas-grekas** [comment]
ping @symfony/deciders

**Comment by dunglas** [comment]
:+1:

File: src/Symfony/Component/Cache/Adapter/ChainAdapter.php [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class ChainAdapter implements AdapterInterface
+{
+    private $adapters = array();
+    private $saveUp;
+
+    /**
+     * @param CacheItemPoolInterface[] $adapters    The ordered list of adapters used to fetch cached items.
+     * @param int                      $maxLifetime The max lifetime of items propagated from lower adapters to upper ones.
```
Comment (fabpot) [inline src/Symfony/Component/Cache/Adapter/ChainAdapter.php]: dots at the end of @param lines should be removed

**Comment by fabpot** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

File: src/Symfony/Component/Cache/Adapter/ChainAdapter.php (hunk 4 above)
Comment (fabpot) [inline src/Symfony/Component/Cache/Adapter/ChainAdapter.php]: b0fffac

**Pull request merged**